### DISCUSSION TITLE
Implementing Data Product API functionality using [a python framework] deployed in Cloud Platform

### DISCUSSION THREAD

**Discussion by tom-webber** [ministryofjustice/analytical-platform#2648] [Firebreak Winter 23]
**Idea Description** - Re-implement some/all of the functionality from the existing DaaP (Data as a Product) using a python framework such as FastAPI, LiteStar, Sanic, Django RF, deployed in Cloud Platform. As with the Lambda Powertools framework, this could be limited to the CRUD metadata/schema endpoints as a proof of concept.

**Why Should We Do This** - We'll get all of the functionality of the existing API, plus all of the integration features of using an API framework. As with the Lambda Powertools proposal, the hope is that this will reduce the size/complexity of the codebase, and help us to move faster in beta

**Definition of Done** - A fully or partially featured API available at a cloud-platform domain

**What skills do we need?** - API Dev, Cloud Platform deployment, API design experience

[Working repo](https://github.com/ministryofjustice/data-platform-firebreak-python-api/tree/main)

**Comment by julialawrence** [comment]
Really cool idea! I was wondering if it's worth removing API Gateway from the equation!

**Comment by MatMoore** [comment]
Even if we don't eliminate API gateway, we could configure it as a [proxy with a greedy path pattern](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html) so that we don't have to repeat all the routing logic there.

Also, we would grant IAM permissions at the granularity of the API service, rather than having to faff around it with every time we add endpoints.

**Comment by michaeljcollinsuk** [comment]
If you are using FastAPI think there will be some crossover with https://github.com/ministryofjustice/data-platform/discussions/2588 around setup, deployment etc so I'd like to be involved in that capacity at least

**Comment by MatMoore** [comment]
I'm interested on collaborating on this too (if there's time I can maybe split time between this one and the powertools one to compare/contrast)

**Comment by mitchdawson1982** [comment]
Interested on collaborating on this one as well as the Michael's Aws role [microservice](https://github.com/ministryofjustice/data-platform/discussions/2588)
